win32/game_of_life/game_of_life.odin

```diff
@@ -0,0 +1,436 @@
+		idx := rand.int31_max(siz)
+		app.world.alive[idx] = 1
+	case '£':
+		show_error_and_panic("panic test")
+	}
+	return 0
+}
+
+wndproc :: proc "system" (hwnd: win32.HWND, msg: win32.UINT, wparam: win32.WPARAM, lparam: win32.LPARAM) -> win32.LRESULT {
+	context = runtime.default_context()
+	switch msg {
+	case win32.WM_CREATE:     return WM_CREATE(hwnd, lparam)
+	case win32.WM_DESTROY:    return WM_DESTROY(hwnd)
+	case win32.WM_ERASEBKGND: return 1 // paint should fill out the client area so no need to erase the background
+	case win32.WM_PAINT:      return WM_PAINT(hwnd)
+	case win32.WM_CHAR:       return WM_CHAR(hwnd, wparam, lparam)
+	case win32.WM_TIMER:      return WM_TIMER(hwnd, wparam, lparam)
+	case:                     return win32.DefWindowProcW(hwnd, msg, wparam, lparam)
+	}
+}
+
+register_class :: proc(instance: win32.HINSTANCE) -> win32.ATOM {
+	icon: win32.HICON = win32.LoadIconW(instance, win32.MAKEINTRESOURCEW(101))
+	if icon == nil {icon = win32.LoadIconW(nil, win32.wstring(win32._IDI_APPLICATION))}
+	if icon == nil {show_error_and_panic("Missing icon")}
+	cursor := win32.LoadCursorW(nil, win32.wstring(win32._IDC_ARROW))
+	if cursor == nil {show_error_and_panic("Missing cursor")}
+	wcx := win32.WNDCLASSW {
+		style         = win32.CS_HREDRAW | win32.CS_VREDRAW | win32.CS_OWNDC,
+		lpfnWndProc   = wndproc,
+		hInstance     = instance,
+		hIcon         = icon,
+		hCursor       = cursor,
+		lpszClassName = L("OdinMainClass"),
+	}
+	return win32.RegisterClassW(&wcx)
+}
+
+unregister_class :: proc(atom: win32.ATOM, instance: win32.HINSTANCE) {
+	if atom == 0 {show_error_and_panic("atom is zero")}
+	if !win32.UnregisterClassW(win32.LPCWSTR(uintptr(atom)), instance) {show_error_and_panic("UnregisterClassW")}
+}
+
+adjust_size_for_style :: proc(size: ^int2, dwStyle: win32.DWORD) {
+	rect := win32.RECT{0, 0, size.x, size.y}
+	if win32.AdjustWindowRect(&rect, dwStyle, false) {
+		size^ = {i32(rect.right - rect.left), i32(rect.bottom - rect.top)}
+	}
+}
+
+center_window :: proc(position: ^int2, size: int2) {
+	if deviceMode: win32.DEVMODEW; win32.EnumDisplaySettingsW(nil, win32.ENUM_CURRENT_SETTINGS, &deviceMode) {
+		device_size := int2{i32(deviceMode.dmPelsWidth), i32(deviceMode.dmPelsHeight)}
+		position^ = (device_size - size) / 2
+	}
+}
+
+create_window :: #force_inline proc(instance: win32.HINSTANCE, atom: win32.ATOM, game: ^Game) -> win32.HWND {
+	if atom == 0 {show_error_and_panic("atom is zero")}
+	style :: win32.WS_OVERLAPPED | win32.WS_CAPTION | win32.WS_SYSMENU
+	size := game.window.size
+	pos := int2{i32(win32.CW_USEDEFAULT), i32(win32.CW_USEDEFAULT)}
+	adjust_size_for_style(&size, style)
+	if .CENTER in game.window.control_flags {
+		center_window(&pos, size)
+	}
+	return win32.CreateWindowW(win32.LPCWSTR(uintptr(atom)), game.window.name, style, pos.x, pos.y, size.x, size.y, nil, nil, instance, game)
+}
+
+message_loop :: proc() -> int {
+	msg: win32.MSG
+	for win32.GetMessageW(&msg, nil, 0, 0) > 0 {
```

Comment:
@laytan I did like the BOOL before, but after reading up I agree with the INT change. If you look up https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getmessagew you would think BOOL is fine until you read the last part :/ i know i should check for -1 here and do something else but bailing out in this demo should be fine :D

Reply:
Yeah unfortunately our booleans can only be in 2 states and the function can return 3, I think it is weird they annotate it as BOOL in the first place but that's probably because C is fine with it.